@mattip 
sorry for wasting your time. There was an error on my side and the tests are failed with the latest PR code due to the insertion of the array shape in the preparation of the formats.

I think there is an issue with how fromrecords calls fromarrays.

Without considering the input array shapes I get an error with this snippet:

```python
arrays = [
    np.arange(10).reshape((5, 2)),
    np.ones(5) * -1,
]

names = ['f1', 'f2']

arr = np.core.records.fromarrays(arrays, names=names)
```
```
1.13.3
Traceback (most recent call last):
  File "/home/daniel/.config/spyder-py3/temp.py", line 25, in <module>
    arr = np.core.records.fromarrays(arrays, names=names)
  File "/usr/lib/python3/dist-packages/numpy/core/records.py", line 622, in fromarrays
    raise ValueError("array-shape mismatch in array %d" % k)
ValueError: array-shape mismatch in array 1

```
I find this unexpected since the input arrays have the same first dimension

If the shape is considered then the snippet runs ok.

--- a/benchmarks/benchmarks/bench_records.py
+++ b/benchmarks/benchmarks/bench_records.py
@@ -1,0 +1,32 @@
+from __future__ import absolute_import, division, print_function
+
+from .common import Benchmark
+
+import numpy as np
+
+        
+class Records(Benchmark):
+    def setup(self):
+        self.l50 = np.arange(1000)
+        self.fields_number = 10000
+        self.arrays = [self.l50 for _ in range(self.fields_number)]
+        self.formats = [self.l50.dtype.str for _ in range(self.fields_number)]
+        self.formats_str = ','.join(self.formats)
+        self.dtype_ = np.dtype(
+            [
+                ('field_{}'.format(i), self.l50.dtype.str)
+                for i in range(self.fields_number)
+            ]
+        )
+        
+    def time_fromarrays_w_dtype(self):
+        np.core.records.fromarrays(self.arrays, dtype=self.dtype_)
+        
+    def time_fromarrays_wo_dtype(self):
+        np.core.records.fromarrays(self.arrays)
+        
+    def time_fromarrays_formats_as_list(self):
+        np.core.records.fromarrays(self.arrays, formats=self.formats)
+        
+    def time_fromarrays_formats_as_string(self):
+        np.core.records.fromarrays(self.arrays, formats=self.formats_str)

--- a/benchmarks/benchmarks/bench_records.py
+++ b/benchmarks/benchmarks/bench_records.py
@@ -1,10 +1,11 @@
 from __future__ import absolute_import, division, print_function
+import os
 
 from .common import Benchmark
 
 import numpy as np
 
-        
+      
 class Records(Benchmark):
     def setup(self):
         self.l50 = np.arange(1000)
@@ -18,6 +19,14 @@ class Records(Benchmark):
                 for i in range(self.fields_number)
             ]
         )
+        self.buffer = self.l50.tostring() * self.fields_number
+        for file in ('test1.bin', 'test2.bin', 'test3.bin'):
+            with open(file, 'w+b') as f:
+                f.write(self.buffer)
+        
+    def teardown(self):
+        for file in ('test1.bin', 'test2.bin', 'test3.bin'):
+            os.remove(file)
         
     def time_fromarrays_w_dtype(self):
         np.core.records.fromarrays(self.arrays, dtype=self.dtype_)
@@ -30,3 +39,21 @@ class Records(Benchmark):
         
     def time_fromarrays_formats_as_string(self):
         np.core.records.fromarrays(self.arrays, formats=self.formats_str)
+        
+    def time_fromstring_w_dtype(self):
+        np.core.records.fromstring(self.buffer, dtype=self.dtype_)
+        
+    def time_fromstring_formats_as_list(self):
+        np.core.records.fromstring(self.buffer, formats=self.formats)
+        
+    def time_fromstring_formats_as_string(self):
+        np.core.records.fromstring(self.buffer, formats=self.formats_str)
+           
+    def time_fromfile_w_dtype(self):
+        np.core.records.fromfile('test1.bin', dtype=self.dtype_)
+        
+    def time_fromfile_formats_as_list(self):
+        np.core.records.fromfile('test2.bin', formats=self.formats)
+        
+    def time_fromfile_formats_as_string(self):
+        np.core.records.fromfile('test3.bin', formats=self.formats_str)

--- a/numpy/core/records.py
+++ b/numpy/core/records.py
@@ -584,7 +584,9 @@ class recarray(ndarray):
 
 def fromarrays(arrayList, dtype=None, shape=None, formats=None,
                names=None, titles=None, aligned=False, byteorder=None):
-    """ create a record array from a (flat) list of arrays
+    """ create a record array from a (flat) list of arrays. 
+    If both `dtype` and `formats` are `None` then the input arrays must have 
+    the same shape.
 
     >>> x1=np.array([1,2,3,4])
     >>> x2=np.array(['a','dd','xyz','12'])
@@ -612,10 +614,7 @@ def fromarrays(arrayList, dtype=None, shape=None, formats=None,
         for obj in arrayList:
             if not isinstance(obj, ndarray):
                 raise ValueError("item in the array list must be an ndarray.")
-            shape_ = obj.shape[1:] or ''
-            # keep the 1.15.4 behaviour for now
-            shape_ = ''
-            formats.append('{}{}'.format(shape_, obj.dtype.str))
+            formats.append(obj.dtype.str)
 
     if dtype is not None:
         descr = sb.dtype(dtype)
@@ -770,7 +769,9 @@ def fromfile(fd, dtype=None, shape=None, offset=0, formats=None,
     (0.5, 10, 'abcde')
     >>> r.shape
     (10,)
-    """
+    """  
+    if dtype is None and formats is None:
+        raise ValueError("Must have dtype= or formats=")
 
     if (shape is None or shape == 0):
         shape = (-1,)

--- a/numpy/core/records.py
+++ b/numpy/core/records.py
@@ -613,6 +613,8 @@ def fromarrays(arrayList, dtype=None, shape=None, formats=None,
             if not isinstance(obj, ndarray):
                 raise ValueError("item in the array list must be an ndarray.")
             shape_ = obj.shape[1:] or ''
+            # keep the 1.15.4 behaviour for now
+            shape_ = ''
             formats.append('{}{}'.format(shape_, obj.dtype.str))
 
     if dtype is not None:

